Add compliance operator enhancement proposal

jhrozek opened the pull request:
An enhancement proposal for compliance-operator

JAORMX commented on enhancements/compliance-operator.md:
This doesn't fully address the compliance needs, nor the compliance checks that some NIST controls require. So... going for the operator or a way to run openscap on a cluster is really needed.

jhrozek replied:
OK, I'm dropping this section completely then.

JAORMX commented on enhancements/compliance-operator.md:
There are security risks that we need to start considering:

* the operator will need to mount the whole filesystem from the host, we need to make sure that mounting it as read-only is fine to lock down the operator's workload.
* If remediations need to be done, something needs to have permissions to create the MachineConfigs or whatever resources need to be created... we need to come up with a safe scheme to do this, so we don't have an all-powerful operator that could be compromised. Any suggestions on this?

jhrozek replied:
@runcom Hi Antonio, do you have any thoughts on the second point?

jhrozek replied:
@runcom still would welcome your thoughts on the remediations, if you have any :)

ashcrow replied:
> If remediations need to be done, something needs to have permissions to create the MachineConfigs or whatever resources need to be created... we need to come up with a safe scheme to do this, so we don't have an all-powerful operator that could be compromised. Any suggestions on this?

I think that if remediation has to be done automatically then access to the MC's would be needed (service account possibly). If the remediation could be done with the compliance operators help then the compliance operator could generate the MC and pass it to the admin in one way or another. Think selinux and audit log.

vrutkovs commented on enhancements/compliance-operator.md:
```diff
@@ -0,0 +1,222 @@
+
+## Proposal
+
+An instance of a scan would be represented by a CRD. The CRD would include
```
Do all nodes get checked using the same set of rules or it can be customized?

JAORMX replied:
It can be customized. There are a couple of ways of doing this (none of them are right or wrong according to the OpenSCAP team). you can create a profile per role (e.g. one profile for masters and one profile for workers), or you can create a general-purpose compliance checks that merely ignores the check if it doesn't apply (e.g. if you need to check for a service that only exists on the master nodes, then you detect that the service isn't there, and ignore the check on the workers).

vrutkovs replied:
Ah, that makes sense. Seems CRD should mention MachineSet it would operate on then

jhrozek replied:
Yeah, probably. Initially I was thinking just nodeSelectors and match against labels, but if a MachineSet makes more sense, then sure, it's an option. (I need to familiarize myself with MachineSets a bit though before I incorporate that into the proposal..)

jhrozek replied:
Hi @vrutkovs would you mind helping me expand on the idea? As far as I cans see in my cluster, the MachineSets only reference worker nodes by default, even though [the documentation](https://docs.openshift.com/container-platform/4.1/machine_management/creating-infrastructure-machinesets.html) suggests to create MachineSets for infrastructure components as well. Also, from the points of view of the compliance operator, you'd typically only want to match against labels, typically `node-role` or perhaps `node.openshift.io/os_id`.

So at the moment it is not clear to me how exactly would you suggest to use MachineSets, or what would that give over using something like:
```
nodeSelector:
   node-role.kubernetes.io/worker=
```

cgwalters commented on enhancements/compliance-operator.md:
```diff
@@ -0,0 +1,222 @@
+* Inspect the scan result on the high level (pass/fail/error)
+* Provide a way to collect the detailed results of the scan
+  * The results would list a way to remediate the gaps, typically by the means
+    of deploying certain manifests to OKD (e.g. a Machineconfig), a different
```
s/OKD/OpenShift/ ?

jhrozek replied:
Probably? :)

Just to get the terminology right, OKD is used for the Kubernetes piece only and OpenShift = OKD + RHCOS?

JAORMX commented:
> I think my biggest question here is: What types of things would OpenSCAP find against an OpenShift cluster? Surely we'd make sure that our default installs are compliant?

That's unfortunately not possible for some cases. For instance, FedRAMP moderate has very strict requirements related to the IdP, and we cannot simply configure a compliant IdP by default.
For RHCOS, there are some auditd rules that are required, that, if enabled, would inpact performance.
> 
> IOW, if a customer's cluster trips something here...it's probably a bad MachineConfig or a custom privileged pod/daemonset mounting the host and doing stuff right?
> 
> If the above is all true, then it feels like it'd be best if we can try to track down what caused it - and remediation isn't creating a _new_ MC, it's fixing the existing one or daemonset.

jhrozek commented:
On 10/25/19 4:53 PM, Colin Walters wrote:
>
> I think my biggest question here is: What types of things would 
> OpenSCAP find against an OpenShift cluster?
>

This depends on what kind of profile you test against and want to 
conform with. We've been mostly looking at FedRAMP moderate, so I'll use 
that as an example below.

> Surely we'd make sure that our default installs are compliant?
>

This might not be doable in the general case. Some checks might be 
something that should be enabled by default, e.g. one thing we'd check 
for in the FedRAMP moderate profile is "is SELinux enforcing?". But 
there are other things that are dependant on the environment. For 
example, one requirement is that there must be a legal notice displayed 
after login for legal reasons and the way you'd do that is by creating a 
ConfigMap named "motd" in the "openshift" namespace. And there the check 
would just check for the existence of that ConfigMap and fail if there 
is none -- but the contents must be filled in by the administrator. Or, 
as another example, the profile might require that certain audit rules 
for the auditd deamon are enabled. But the rules have performance 
impact, so it's probably not a good idea to enable them by default for 
all installations. etc.

> IOW, if a customer's cluster trips something here...it's probably a 
> bad MachineConfig or a custom privileged pod/daemonset mounting the 
> host and doing stuff right?
>
> If the above is all true, then it feels like it'd be best if we can 
> try to track down what caused it - and remediation isn't creating a 
> /new/ MC, it's fixing the existing one or daemonset.
>

jhrozek commented:
I pushed a v2 of the proposal. Changes include:
 - s/OKD/OpenShift/
 - the content would be delivered through a separate image via an `initContainer` that would populate a volume to be consumed by the scanner image
    - CRD adjustted accordingly
    - added a User Story describing why we need this
 - reworded the Risks and Mitigations section
 - dropped the "Alternatives" paragraph as it was not really a feasible alternative

jhrozek commented on enhancements/compliance-operator.md:
Just highlighting this in case anyone thinks this is too hacky and we should do things differently.

jhrozek replied:
(Sorry for talking to myself)
One thing that was proposed to me earlier in a discussion was to use secrets instead, because the compliance scan results might be security sensitive and at least the intent of secrets aligns better with that kind of information.

JAORMX commented on enhancements/compliance-operator.md:
```diff
@@ -0,0 +1,246 @@
+
+This is a use-case where the admin just wants to visualize the gaps in
+compliance. They would perform the remediation themselves, potentially
+based on the MCOs the compliance operator suggests.
```
s/MCOs/MachineConfigs/

MCO stands for Machine-Config-Operator, so that's not what we're trying to portray here.

ashcrow commented on enhancements/compliance-operator.md:
How about:

A CI test that
- Rolls out the operator
- Ensures the operator is running
- Executes a scan
- Ensures a result is produced and the scheduled container is cleaned up

jhrozek replied:
Sounds good, I worded the paragraph a little differently, but I think it's still quite close to what you suggested.

ashcrow commented:
/cc @lucab @miabbott

ashcrow commented:
There are a few missing sections still. One is only required when targeted (so it's fine). Is there a plan to have the other sections updated before final review?

jhrozek commented:
On 10/31/19 5:00 PM, Stephen Milner wrote:
>
> There are a few missing sections still. One is only required when 
> targeted (so it's fine). Is there a plan to have the other sections 
> updated before final review?
>
Thank you, I added all the missing sections except the one required only 
when targeting a release.

I also explicitly mentioned that the operator must be able to read the 
resources it would be checking and listed all the reviewers so far.

ashcrow commented:
Thank you @jhrozek. The changes look great!

ashcrow commented:
Unless anyone objects this looks ready to merge to me. I'll give it a few days.

ashcrow commented:
/lgtm
